### PR TITLE
Cascading binding deletion

### PR DESCRIPTION
**Pull request by piotrmiskiewicz**
**Implements proposal: https://github.com/kubernetes-sigs/service-catalog/issues/2734**

Adds a cascading deletion feature which deletes existing service bindings for a service instance which is being deleted.

**Which issue(s) this PR fixes** 
Fixes:  https://github.com/kubernetes-sigs/service-catalog/issues/481 https://github.com/kyma-project/kyma/issues/5180

### CODE REVIEW

**Comment by piotrmiskiewicz** [comment]
How to test:

Create a namespace:
```kubectl create ns test-ns```

Create an instance
```
cat <<EOF | kubectl apply -f -
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: test-instance
  namespace: test-ns
spec:
  clusterServiceClassExternalName: test-service
  clusterServicePlanExternalName: default
EOF
```
Create 3 bindings:
```
cat <<EOF | kubectl apply -f -
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: test-binding1
  namespace: test-ns
spec:
  instanceRef:
    name: test-instance
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: test-binding2
  namespace: test-ns
spec:
  instanceRef:
    name: test-instance
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: test-binding3
  namespace: test-ns
spec:
  instanceRef:
    name: test-instance     
EOF
```

**Comment by piotrmiskiewicz** [comment]
The PR implements service bindings deletion in the controller due to a business requirement - deprovision must be done after all unbind operations.
The idea of having ownerReference in the ServiceBinding instances and use K8s garbage collector does not work. The k8s garbage collector will try to delete bindings, when service instance is deleted, but it can be deleted, when all unbind operation is done.

**Comment by MHBauer** [comment]
OMG, this is part of the matrix of deleting things! #2229

**Comment by mszostok** [comment]
I still want to discuss that in details 
- why `kubectl delete .. --cascade=true` cannot be used? Maybe we can reuse `"propagationPolicy":"Foreground"` concept but implement it by our own?
- or why the spec was not changed? We should have a new field under ServiceInstance spec, sth like deletionPolicy. By default, it will have old behavior. If a user will change that then the cascading opt will be used. 

The current solution is not backward compatible, so let's discuss that if we want to implement this feature in that way.

/hold

**Comment by piotrmiskiewicz** [comment]
There is a feature flag which enables cascading deletion, so the change is backward compatible. The ownerReference cannot be used because of a problem when the delete is done with propagationPolicy=background. The binding must be deleted before the instance. The risk is someone could do `kubectl delete serviceinstance ...` without the `--cascade=true` flag

**Comment by mszostok** [comment]
> Feature gates are a set of key=value pairs that describe alpha or experimental features and can be turned on or off

so it's not backward compatible because you cannot remove that feature flag and make that behavior as the core one because u will change the deletion policy for all clients when they will update to new service catalog version. 

I'm talking about the `"propagationPolicy":"Foreground"` not the `ownerReference`. We can impl that GC on our own as we did here but based on `propagationPolicy` but I'm not fully convinced to do that in that way. It's just an option and maybe better will be to extend the ServiceInstance `Spec`. IMO it will be also worth to show those two approaches in sig-api-machinery and ask them about the recommendation.

let's discuss that offline and post here clear strategy for having that feature in ServiceCatalog:)

File: docs/feature-gates.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ different Service Catalog.
 | `ResponseSchema` | `false` | Alpha | v0.1.12 | |
 | `ServicePlanDefaults` | `false` | Alpha | v0.1.32 | |
 | `UpdateDashboardURL` | `false` | Alpha | v0.1.13 | |
+| `CascadingDeletion` | ` false` | Alpha | v0.2.3 | |
```
Comment (mszostok) [inline docs/feature-gates.md]:
```suggestion
| `CascadingDeletion` | ` false` | Alpha | v0.3.0 | |
```

File: pkg/controller/controller_instance.go [hunk 2]
```diff
@@ -2218,6 +2234,46 @@ func (c *controller) checkServiceInstanceHasExistingBindings(instance *v1beta1.S
 	return nil
 }
 
+func (c *controller) listExistingBindings(instance *v1beta1.ServiceInstance) ([]*v1beta1.ServiceBinding, error) {
+	bindingLister := c.bindingLister.ServiceBindings(instance.Namespace)
+
+	selector := labels.NewSelector()
+	bindingList, err := bindingLister.List(selector)
```
Comment (mszostok) [inline pkg/controller/controller_instance.go]: just inline all of those statements:
```go
c.bindingLister.ServiceBindings(instance.Namespace).List(labels.NewSelector())
```

File: pkg/controller/controller_instance.go [hunk 3]
```diff
@@ -2218,6 +2234,46 @@ func (c *controller) checkServiceInstanceHasExistingBindings(instance *v1beta1.S
...
+	}
+
+	return found, nil
+}
+
+func (c *controller) deleteExistingBindings(instance *v1beta1.ServiceInstance) error {
+	klog.V(4).Infof("Delete existing bindings for the instance %s", instance.Name)
+	bindings, err := c.listExistingBindings(instance)
+	if err != nil {
+		return err
```
Comment (mszostok) [inline pkg/controller/controller_instance.go]: u're missing context for each error
e.g. here u should have 

```go
errors.Wrapf(err, "while fetching ServiceBindings for instance %s", instance.Name) 
```

**Comment by mszostok** [comment]
also, adjust the chart with the new param that can be set by our client

File: pkg/controller/controller_instance.go (hunk 3 above)
Comment (piotrmiskiewicz) [reply pkg/controller/controller_instance.go]: I've added wrapping errors

**Comment by mszostok** [comment]
@piotrmiskiewicz the Helm Chart is still not adjusted, please add opt to enable it, sth similar what we have already: https://github.com/kubernetes-sigs/service-catalog/blob/master/charts/catalog/values.yaml#L105-L106

**Comment by mszostok** [comment]
just add `disable-cascading-deletion` flag beside that  
/lgtm

File: charts/catalog/values.yaml [hunk 4]
```diff
@@ -82,6 +82,8 @@ controllerManager:
   serviceAccount: service-catalog-controller-manager
   # Whether the controller will expose metrics on /metrics
   enablePrometheusScrape: false
+  # Whether the cascading binding deletion is disabled, it overrides the CascadingDeletion feature gate
```
Comment (jberkhahn) [inline charts/catalog/values.yaml]: why the double feature gate?

**Comment by jberkhahn** [comment]
so, i can't seem to get this to work. I set the "cascadingDeletionEnabled" flag to true, and leave the "cascadingDeletionDisabled" to false, as it's a negative, which should make this work? these flags are confusing. why do we not just have a single top level flag like the other feature gates?

**Comment by piotrmiskiewicz** [comment]
As I understand the process, the feature gate will be removed in the future and the feature will be enabled by default. The second (disabling) flag is designed for all users which does not want this feature and it should be set explicitly. Long term the feature gate will be removed and the cascading deletion will be enabled by default. There will be only the cascadingDeletionDisabled flag.

Currently, the cascading deletion feature is an alpha feature. This is the reason to use a feature gate.
I understand the names in the values.yaml could be confusing: cascadingDeletionEnabled and cascadingDeletionDisabled. If you have better ideas I can change it.

**Comment by piotrmiskiewicz** [comment]
I've made a mistake in the chart, which I've just fixed.

**Comment by jberkhahn** [comment]
why do we not leave the to-be-used-in-the-future-flag out of the chart now, and add it it later, then? I also think the double negative is weird, wouldn't it make more sense to make it --enable-foobar and set it to true by default?

**Comment by piotrmiskiewicz** [comment]
I understand that two options may be confusing, at the beginning I had only the feature gate.
On the other hand, the whole feature "cascading binding deletion" is an alpha feature and needs to be explicitly enabled (by default disabled). After some time - it can be enabled by default. 
Are you sure we can change the way how the controller works just with the next release? No "alpha" stage for few releases?

**Comment by jberkhahn** [comment]
you can leave the option on the actual controller itself, although I would prefer it was a simple "enable-feature-foobar" flag rather than a weird double-negative one. I'm just saying take the ability to set it out of the helm charts so people don't get confused about why there's two settings that look like they do the same thing.

**Comment by piotrmiskiewicz** [comment]
We want to have the cascading deletion feature enabled by default in the future. 
Do you mean something like: `--enable-cascading-deletion` which is enabled by default.
I someone doesn't want to use this feature, he could pass: `--enable-cascading-deletion=false`

Sounds good, but the problem is we are adding such feature without an alpha stage and this feature breaks the backward compatibility.

**Comment by piotrmiskiewicz** [comment]
I removed the "disabling" flag. Now, there is only enabling flag

**Pull request merged**